src/bin/genemichaels.rs

```diff
@@ -250,14 +351,149 @@ fn main() {
                 Ok(_) => { },
                 Err(e) => {
                     if !args.quiet {
-                        eprintln!("Error formatting {}: {:?}", &file.to_string_lossy(), e);
+                        print_error_text();
+                        eprintln!("formatting {}: {:?}", &file.to_string_lossy(), e);
                     }
                     failed = true;
                 },
             };
         }
+        eprintln!(
+            "\x1B[1;32m    Finished\x1B[0;22m workspace formatted successfully in {:.2}s",
+            time::Instant::now().duration_since(inst).as_secs_f64()
+        );
         if failed {
             process::exit(1);
         }
     }
 }
+
+fn process_cargo_toml(path: PathBuf, manifest: Manifest, config: FormatConfig) {
+    let mut dirs = manifest.workspace.map(|wkspc| wkspc.members).map(|mb| {
+        let mut rust_dirs: Vec<PathBuf> =
+            mb.into_iter().filter_map(|m| path.join(&m).exists().then(|| path.join(m))).collect();
+        if let Some(bins) = manifest.bin {
+            for bin in bins {
+                rust_dirs.push(
+                    match bin
+                        .path
+                        .and_then(|p| p.parse::<PathBuf>().ok())
+                        .and_then(|p| p.parent().map(|pp| pp.to_path_buf())) {
+                        Some(p) => p,
+                        None => {
+                            // default location
+                            if path.join("bin").exists() {
+                                rust_dirs.push(path.join("bin"));
+                            }
+                            continue;
+                        },
+                    },
+                )
+            }
+        }
+        if let Some(lib) = manifest.lib {
+            if let Some(p) =
+                lib
+                    .path
+                    .and_then(|p| p.parse::<PathBuf>().ok())
+                    .and_then(|p| p.parent().map(|pp| pp.to_path_buf())) {
+                rust_dirs.push(p)
+            }
+        }
+        if let Some(benches) = manifest.bench {
+            for bench in benches {
+                rust_dirs.push(
+                    match bench
+                        .path
+                        .and_then(|p| p.parse::<PathBuf>().ok())
+                        .and_then(|p| p.parent().map(|pp| pp.to_path_buf())) {
+                        Some(p) => p,
+                        None => {
+                            // default location
+                            if path.join("benches").exists() {
+                                rust_dirs.push(path.join("benches"));
+                            }
+                            continue;
+                        },
+                    },
+                )
+            }
+        }
+        if let Some(tests) = manifest.test {
```

Comment:
I think you can simplify these like `for test in manifest.test.iter().flatten()`, gets rid of some extra logic.

Reply:
That's neat! Cheers